File: backend/app/routers/scenarios.py
from __future__ import annotations

from fastapi import APIRouter, Request
from fastapi.concurrency import run_in_threadpool

from ..core.db import get_settings
from ..core.errors import ApiError
from ..services import repo
from ..services.analytics import aggregate_counterfactuals
from ..services.counterfactual import compute_counterfactual

router = APIRouter()


@router.get("")
async def list_scenarios() -> list[dict]:
    return repo.list_scenarios()


@router.get("/{scenario_id}")
async def get_scenario(scenario_id: int) -> dict:
    s = repo.get_scenario(scenario_id)
    if not s:
        raise ApiError("Not found", 404)
    return s


@router.post("")
async def create_scenario(request: Request) -> dict:
    body = await request.json() or {}
    if not body.get("name") or not body.get("config"):
        raise ApiError("name and config required", 400)
    return repo.insert_scenario(body["name"], body["config"])


@router.put("/{scenario_id}")
async def update_scenario(scenario_id: int, request: Request) -> dict:
    body = await request.json() or {}
    s = repo.update_scenario(scenario_id, body.get("name"), body.get("config"))
    if not s:
        raise ApiError("Not found", 404)
    return s


@router.delete("/{scenario_id}")
async def delete_scenario(scenario_id: int) -> dict:
    repo.delete_scenario(scenario_id)
    return {"ok": True}


@router.post("/run")
async def run(request: Request) -> dict:
    config = await request.json() or {}
    return await run_in_threadpool(run_scenario, config)


def run_scenario(config: dict) -> dict:
    settings = get_settings()
    benchmark = config.get("benchmarkSymbol") or settings["defaultBenchmarkSymbol"]
    pre_tax = config.get("preTax") or False

    ids = config.get("includedInstrumentIds") or []
    if config.get("sellAllToEtf") or len(ids) == 0:
        ids = [i["id"] for i in repo.list_instruments()]

    from_date = config.get("fromDate") or None
    as_of = config.get("asOf") or None

    per_position = []
    for id_ in ids:
        inst = repo.get_instrument(id_)
        if not inst:
            continue
        txs = repo.get_transactions(id_)
        if from_date:
            txs = [t for t in txs if t["date"] >= from_date]
        if not txs:
            continue
        cf = compute_counterfactual(inst, txs, benchmark, settings, pre_tax, as_of=as_of)
        per_position.append({"instrumentId": id_, "symbol": inst["symbol"], "counterfactual": cf})

    aggregate = aggregate_counterfactuals([p["counterfactual"] for p in per_position])
    return {
        "scenario": {**config, "benchmarkSymbol": benchmark, "preTax": pre_tax},
        "perPosition": per_position,
        "aggregate": {
            "actualValueCHF": aggregate["actualValueCHF"],
            "counterfactualValueCHF": aggregate["counterfactualValueCHF"],
            "deltaCHF": aggregate["deltaCHF"],
            "deltaPct": aggregate["deltaPct"],
            "series": aggregate["series"],
        },
    }

File: backend/src/routes/scenarios.ts
import { Router } from 'express';
import type { ScenarioConfig, ScenarioResult } from '@decisionguru/shared';
import { getSettings } from '../db/index.js';
import {
  listScenarios,
  getScenario,
  insertScenario,
  updateScenario,
  deleteScenario,
  getInstrument,
  getTransactions,
  listInstruments,
} from '../services/repo.js';
import { computeCounterfactual } from '../services/counterfactual.js';
import { aggregateCounterfactuals } from './analysis.js';

export const scenariosRouter = Router();

scenariosRouter.get('/', (_req, res) => res.json(listScenarios()));

scenariosRouter.get('/:id', (req, res) => {
  const s = getScenario(Number(req.params.id));
  if (!s) return res.status(404).json({ error: 'Not found' });
  res.json(s);
});

scenariosRouter.post('/', (req, res) => {
  const { name, config } = req.body ?? {};
  if (!name || !config) return res.status(400).json({ error: 'name and config required' });
  res.json(insertScenario(name, config));
});

scenariosRouter.put('/:id', (req, res) => {
  const { name, config } = req.body ?? {};
  const s = updateScenario(Number(req.params.id), name, config);
  if (!s) return res.status(404).json({ error: 'Not found' });
  res.json(s);
});

scenariosRouter.delete('/:id', (req, res) => {
  deleteScenario(Number(req.params.id));
  res.json({ ok: true });
});

/** Run an ad-hoc config (not necessarily saved). */
scenariosRouter.post('/run', async (req, res) => {
  const config = req.body as ScenarioConfig;
  const result = await runScenario(config);
  res.json(result);
});

export async function runScenario(config: ScenarioConfig): Promise<ScenarioResult> {
  const settings = getSettings();
  const benchmark = config.benchmarkSymbol || settings.defaultBenchmarkSymbol;
  const preTax = config.preTax ?? false;

  let ids = config.includedInstrumentIds ?? [];
  if (config.sellAllToEtf || ids.length === 0) {
    ids = listInstruments().map((i) => i.id);
  }

  const perPosition: ScenarioResult['perPosition'] = [];
  for (const id of ids) {
    const inst = getInstrument(id);
    if (!inst) continue;
    const txs = getTransactions(id);
    if (!txs.length) continue;
    const cf = await computeCounterfactual(inst, txs, benchmark, settings, preTax);
    perPosition.push({ instrumentId: id, symbol: inst.symbol, counterfactual: cf });
  }

  const aggregate = aggregateCounterfactuals(perPosition.map((p) => p.counterfactual));
  return {
    scenario: { ...config, benchmarkSymbol: benchmark, preTax },
    perPosition,
    aggregate: {
      actualValueCHF: aggregate.actualValueCHF,
      counterfactualValueCHF: aggregate.counterfactualValueCHF,
      deltaCHF: aggregate.deltaCHF,
      deltaPct: aggregate.deltaPct,
      series: aggregate.series,
    },
  };
}
